content/blog/ruby-boolean.md

```diff
@@ -0,0 +1,128 @@
+
+```ruby
+legion501 = stormtroopers.select{ |trooper| trooper.alive and trooper.commander == "Dark Lord" }
+
+```
+
+Well, everything looks fine here. Till you click on `alive` in your IDE and are taken to its definition. You find out that some newbie trooper has added this code (without tests (╥﹏╥)) and instead of assigning a boolean value, they have assigned it integer values. "Well, big deal", you think assuming that `0` implies `false` and `1` implies `true` always. With that assumption, you begin testing.
+
+```ruby
+irb> StormTrooper.all.to_a.select{ |trooper| not trooper.alive }
+
+```
+Only to get all the stormtroopers. "Surely, not so many can still be alive", you ponder as you begin doubting the library implementation now.
+
+```ruby
+irb> 0 ? "true" : "false"
+=> "true" # Hmm..it must be reversed in ruby. 1 must be false now
+
+irb> 1 : "true" : "false"
+=> "true"
+
+irb> what the heck ruby
+(irb):in `<main>': undefined local variable or method `ruby' for main (NameError)
+
+```
+
+Not knowing what is happening, you go ahead and change `alive` to take boolean values - `true` and `false` and sure enough, that solves the issue. You avert the risk on your life and become increasingly curious about this.
+
+> **What is the point of 0 or 1 or any number to be accepted for a boolean parameter if everything is going to be evaluated to `true`?**  
+> Why is only `true` and `false` actually `true` and `false` respectively?
+
+In other higher languages such as Python, `0` is treated as `False` and `1` as `True` because `bool` is a subclass of `int`. So it's not that the integers are evaluated to boolean values but integers have the boolean value.
+
+```python
+>>> False == 0
+True
+>>> False == 1
+False
+>>> True == 1
+True
+>>> True == 0
+False
+
+```
+All of this makes perfect sense. Then why not Ruby?
+
+```ruby
+irb> 0.class.ancestors
+=> [Integer, Numeric, Comparable, Object, PP::ObjectMixin, Kernel, BasicObject]
+irb> true.class.ancestors
+=> [TrueClass, Object, PP::ObjectMixin, Kernel, BasicObject]
+
+```
+Weirdly, there are no common ancestors between integer class and boole..? No boolean class here?!
+
+```ruby
+irb> false.class.ancestors
+=> [FalseClass, Object, PP::ObjectMixin, Kernel, BasicObject]
+
+```
+Not only there is no `Boolean` class but `true` and `false` have their own classes!?
+
+> Don't `true` and `false` have so much in common? It should make sense to keep them under the same class
+
+```ruby
+irb> false.methods.count
+59
+irb> true.methods.count
+59
+
+# Aha, language fault possibly
+
+irb> (false.methods & true.methods).count
+59
+irb> (false.methods & true.methods) - Object.methods
+=> [:&, :|, :^]
+
+# Most definitely
+
+```
+
+> They only have three methods precisely in common with each other (╭ರ_•́)
```

Comment:
(╭ರ_•́)

Maybe we can use the monocles emoji here 🧐 some people may get confused by this

Reply:
Didn't want to add a hint of yellow or color in a blog, so went with ascii emojis